sys-utils/setpriv-landlock.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Copyright (C) 2023 Thomas Weißschuh <thomas@t-8ch.de>
+ */
+
+#include <sys/prctl.h>
+#include <sys/syscall.h>
+#include <linux/landlock.h>
+
+#include "setpriv-landlock.h"
+
+#include "strutils.h"
+#include "xalloc.h"
+#include "nls.h"
+#include "c.h"
+
+#ifndef HAVE_LANDLOCK_CREATE_RULESET
+static inline int landlock_create_ruleset(
+		const struct landlock_ruleset_attr *attr,
+		size_t size, uint32_t flags)
+{
+	return syscall(__NR_landlock_create_ruleset, attr, size, flags);
+}
+#endif
+
+#ifndef HAVE_LANDLOCK_ADD_RULE
+static inline int landlock_add_rule(
+		int ruleset_fd, enum landlock_rule_type rule_type,
+		const void *rule_attr, uint32_t flags)
+{
+	return syscall(__NR_landlock_add_rule, ruleset_fd, rule_type,
+		       rule_attr, flags);
+}
+#endif
+
+#ifndef HAVE_LANDLOCK_RESTRICT_SELF
+static inline int landlock_restrict_self(int ruleset_fd, uint32_t flags)
+{
+	return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#define SETPRIV_EXIT_PRIVERR 127	/* how we exit when we fail to set privs */
+
+struct landlock_rule_entry {
+	struct list_head head;
+	enum landlock_rule_type rule_type;
+	union {
+		struct landlock_path_beneath_attr path_beneath_attr;
+	};
+};
+
+static const struct {
+	unsigned long long value;
+	const char *type;
+} landlock_access_fs[] = {
+	{ LANDLOCK_ACCESS_FS_EXECUTE,     "execute"     },
+	{ LANDLOCK_ACCESS_FS_WRITE_FILE,  "write-file"  },
+	{ LANDLOCK_ACCESS_FS_READ_FILE,   "read-file"   },
+	{ LANDLOCK_ACCESS_FS_READ_DIR,    "read-dir"    },
+	{ LANDLOCK_ACCESS_FS_REMOVE_DIR,  "remove-dir"  },
+	{ LANDLOCK_ACCESS_FS_REMOVE_FILE, "remove-file" },
+	{ LANDLOCK_ACCESS_FS_MAKE_CHAR,   "make-char"   },
+	{ LANDLOCK_ACCESS_FS_MAKE_DIR,    "make-dir"    },
+	{ LANDLOCK_ACCESS_FS_MAKE_REG,    "make-reg"    },
+	{ LANDLOCK_ACCESS_FS_MAKE_SOCK,   "make-sock"   },
+	{ LANDLOCK_ACCESS_FS_MAKE_FIFO,   "make-fifo"   },
+	{ LANDLOCK_ACCESS_FS_MAKE_BLOCK,  "make-block"  },
+	{ LANDLOCK_ACCESS_FS_MAKE_SYM,    "make-sym"    },
+#ifdef LANDLOCK_ACCESS_FS_REFER
+	{ LANDLOCK_ACCESS_FS_REFER,       "refer"       },
+#endif
+#ifdef LANDLOCK_ACCESS_FS_TRUNCATE
+	{ LANDLOCK_ACCESS_FS_TRUNCATE,    "truncate"    },
+#endif
+};
+
+static long landlock_access_to_mask(const char *str, size_t len)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(landlock_access_fs); i++)
+		if (strncmp(landlock_access_fs[i].type, str, len) == 0)
+			return landlock_access_fs[i].value;
+	return -1;
+}
+
+static uint64_t parse_landlock_fs_access(const char *list)
+{
+	unsigned long r = 0;
+	size_t i;
+
+	/* without argument, match all */
+	if (list[0] == '\0') {
```

Comment:
`setpriv` is a fairly lowlevel tool. So in my opinion it would be fine to force users to specify explicitly what they want if they need longterm deterministic behavior.
As a default I would like to keep the API as close to the kernel as possible, providing users with the possibility to discover which rules are supported by the kernel and util-linux via exitcodes.
(I see that this also needs to be reworked)

This should be better documented.

Then on top of the base functionality we could add userfriendly extensions.